ethers-solc/src/compile/project.rs

```diff
@@ -0,0 +1,181 @@
+//! Manages compiling of a `Project`
+
+use crate::{
+    error::Result, resolver::GraphEdges, utils, ArtifactOutput, Graph, Project, ProjectPathsConfig,
+    SolFilesCache, Solc, SolcConfig, Source, Sources,
+};
+use std::{
+    collections::{hash_map, BTreeMap, HashMap},
+    path::{Path, PathBuf},
+};
+
+#[derive(Debug)]
+pub struct ProjectCompiler<'a, T: ArtifactOutput> {
+    /// Contains the relationship of the source files and their imports
+    edges: GraphEdges,
+    project: &'a Project<T>,
+    /// how to compile all the sources
+    sources: CompilerSources,
+}
+
+impl<'a, T: ArtifactOutput> ProjectCompiler<'a, T> {
+    pub fn new(project: &'a Project<T>) -> Result<Self> {
+        Self::with_sources(project, project.paths.read_input_files()?)
+    }
+
+    pub fn with_sources(project: &'a Project<T>, sources: Sources) -> Result<Self> {
+        let graph = Graph::resolve_sources(&project.paths, sources)?;
+        // TODO this should return a type that still knows the relationships edges and nodes
+        let (versions, edges) = graph.into_sources_by_version(!project.auto_detect)?;
+
+        let sources_by_version = versions.get(&project.allowed_lib_paths)?;
+
+        let mode = if project.solc_jobs > 1 && sources_by_version.len() > 1 {
+            // if there are multiple different versions and we can use multiple jobs we can compile
+            // them in parallel
+            CompilerSources::Para(sources_by_version, project.solc_jobs)
+        } else {
+            CompilerSources::Sequ(sources_by_version)
+        };
+        Ok(Self { edges, project, sources: mode })
+    }
+
+    /// Compiles all the sources
+    pub fn compile(self) {
+        let Self { edges: _, project: _, sources: _mode } = self;
+
+        todo!()
+    }
+}
+
+/// Determines how the `solc <-> sources` pairs are executed
+#[derive(Debug)]
+enum CompilerSources {
+    /// Compile all these sequentially
+    Sequ(BTreeMap<Solc, Sources>),
```

Comment:
let's do full naming here

```suggestion
    Sequential(BTreeMap<Solc, Sources>),
```